#  Coursara Capstone  - Toronto Neighborhoods

## This project will compare neighborhoods in the Toronto area

By Brian Hoyler

## Part 1

create datatable with Toronto Neighborhoods

In [88]:
import pandas as pd
import numpy as np
print('Libraries imported.')

Libraries imported.


### Install the requirements for BeautifulSoup

In [89]:
#!pip install bs4
#!pip install lxml
#!pip install html5lib
from bs4 import BeautifulSoup # this module helps in web scrapping.
import requests  # this module helps us to download a web page
print("BeautifulSoup installed")

BeautifulSoup installed


### define the URL for the data and use BeautifulSoup to scrape the data

In [90]:
#The below url contains html tables with data about Toronto Neighborhoods.
url = "https://en.wikipedia.org/wiki/List_of_neighbourhoods_in_Toronto"
url2 = "https://en.wikipedia.org/wiki/Demographics_of_Toronto_neighbourhoods"
url3 = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [91]:
# get the contents of the webpage in text format and store in a variable called data
data  = requests.get(url3).text

In [92]:
#soup = BeautifulSoup(data,"lxml")
soup = BeautifulSoup(data,"html5lib")

In [93]:
#find all html tables in the web page
tables = soup.find_all('table') # in html table is represented by the tag <table>

### The following code produces the columns for the dataframe and drops rows with unassigned boroughs

In [94]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)
#print(table_contents)

### build dataframe and modify some of the borough names

In [95]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

## Results of Step One - Creating a dataframe of postalcodes, Boroughs and Neighborhoods

In [96]:
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [97]:
df.shape

(103, 3)

## First portion is complete.  The dataframe has 103 rows.

# Beginning of the second section

### define the geocoder function

In [98]:
import geocoder # import geocoder

def get_coordinates():
# initialize your variable to None
    lat_lng_coords = None

# loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format('M5A'))
        lat_lng_coords = g.latlng
    return lat_lng_coords



### determine the latitude and longitude for each postal code

In [37]:
#lat_lng_coords = get_coordinates()
#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]
#print(latitude,longitude)

### read the csv file

In [13]:
latlongdf = pd.read_csv('Geospatial_Coordinates.csv')

In [99]:
latlongdf.rename(columns = {'Postal Code':'PostalCode'}, inplace = True) 


### now merge the postal code dataframe with the latitude and longitude dataframe.

In [100]:
mergeddf = pd.merge(df, latlongdf, on="PostalCode")

In [102]:
mergeddf

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [103]:
mergeddf.shape

(103, 5)

## Shape of complete data frame is 103 rows by 5 columns

# End of the Second Section

# Begining of the final section - comparing Neighborhoods

### explore the dataframe

In [104]:
toronto_metro = mergeddf
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_metro['Borough'].unique()),
        toronto_metro.shape[0]
    )
)

The dataframe has 15 boroughs and 103 neighborhoods.


## import appropriate libraries

In [105]:
#install geopy 

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

##!pip install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## get latitute and Longintude of Toronto, ON Canada

In [106]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


## Draw map of the Toronto metro area - shows all of the neighborhoods by postal code

In [48]:
# create map of toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_metro['Latitude'], toronto_metro['Longitude'], toronto_metro['Borough'], toronto_metro['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

set up Foursquare credentials

In [240]:
CLIENT_ID = 'GJY3L3YX0HUR2YK0Z24BQD1ANHM1W4OUX3HF2FPCDZKNGAXD' # your Foursquare ID
CLIENT_SECRET = 'E5N3O3LN5UWFBOKUQYCYVF0GFBCNGNDZH0F5VDNEJJVQII1N' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value


ACCESS_TOKEN = 'NZETSIDJQ2C2LMSXZYRSJ5ZYSXJ23OQGIDVQSO0LTUS1WUU5'
radius = 500


In [238]:
## Set up url for foursquare query ov venues

In [239]:
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)


Build functions for use in analysis.
    First function gets the category of the venues
    Next function gets the venues around the neighboorhoods

In [195]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [196]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

## Execute venues function by neighborhood

In [197]:
toronto_venues = getNearbyVenues(names=toronto_metro['Neighborhood'],
                                   latitudes=toronto_metro['Latitude'],
                                   longitudes=toronto_metro['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

## Explore the venues

In [198]:
print(toronto_venues.shape)
toronto_venues.head()

(2126, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3         Parkwoods              43.753259              -79.329656   
4  Victoria Village              43.725882              -79.315572   

                               Venue  Venue Latitude  Venue Longitude  \
0                    Brookbanks Park       43.751976       -79.332140   
1                    Brookbanks Pool       43.751389       -79.332184   
2                      Variety Store       43.751974       -79.333114   
3  Corrosion Service Company Limited       43.752432       -79.334661   
4             Victoria Village Arena       43.723481       -79.315635   

               Venue Category  
0                        Park  
1                        Pool  
2           Food & Drink Shop  
3  Construction & Landscaping  
4                Hockey Arena

In [200]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               8   
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   24   
...                                                                ...   
Willowdale South                                                    33   
Willowdale West                                                      5   
Woburn                                                               4   
Woodbine Heights                                                     6   
York Mills West                                                      2   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    24   
...                                                                 ...   
Willowdale South                                                     33   
Willowdale West                                                       5   
Woburn                                                                4   
Woodbine Heights                                                      6   
York Mills West                                                       2   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   24              24   
...                                                ...             ...   
Willowdale South                                    33              33   
Willowdale West                                      5               5   
Woburn                                               4               4   
Woodbine Heights                                     6               6   
York Mills West                                      2               2   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             24   
...                                                          ...   
Willowdale South                                              33   
Willowdale West                                                5   
Woburn                                                         4   
Woodbine Heights                                               6   
York Mills West                                                2   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        8 

In [201]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 273 unique categories.


## Onehot encoding of the venues to build dataframe for analysis

In [202]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0             0  ...           0                              0   
1             0  ...           0                              0   
2             0  ...           0                              0   
3             0  ...           0                              0   
4             0  ...           0                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0         0          0            0              0  
1         0          0            0              0  
2         0          0            0              0  
3         0          0            0              0  
4         0          0            0              0  

[5 rows x 273 columns]

In [203]:
toronto_onehot.shape

(2126, 273)

In [204]:
toronto_onehot['Neighborhood']

0                                               Parkwoods
1                                               Parkwoods
2                                               Parkwoods
3                                               Parkwoods
4                                        Victoria Village
                              ...                        
2121    Mimico NW, The Queensway West, South of Bloor,...
2122    Mimico NW, The Queensway West, South of Bloor,...
2123    Mimico NW, The Queensway West, South of Bloor,...
2124    Mimico NW, The Queensway West, South of Bloor,...
2125    Mimico NW, The Queensway West, South of Bloor,...
Name: Neighborhood, Length: 2126, dtype: object

In [205]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
94                                 Willowdale South          0.0   
95                                  Willowdale West          0.0   
96                                           Woburn          0.0   
97                                 Woodbine Heights          0.0   
98                                  York Mills West          0.0   

    Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                 0.0      0.0                 0.0           0.0   
1                 0.0      0.0                 0.0           0.0   
2                 0.0      0.0                 0.0           0.0   
3                 0.0      0.0                 0.0           0.0   
4                 0.0      0.0                 0.0           0.0   
..                ...      ...                 ...           ...   
94                0.0      0.0                 0.0           0.0   
95                0.0      0.0                 0.0           0.0   
96                0.0      0.0                 0.0           0.0   
97                0.0      0.0                 0.0           0.0   
98                0.0      0.0                 0.0           0.0   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0              0.0              0.0               0.0             0.000000   
1              0.0              0.0               0.0             0.000000   
2              0.0              0.0               0.0             0.000000   
3              0.0              0.0               0.0             0.000000   
4              0.0              0.0               0.0             0.041667   
..             ...              ...               ...                  ...   
94             0.0              0.0               0.0             0.000000   
95             0.0              0.0               0.0             0.000000   
96             0.0              0.0               0.0             0.000000   
97             0.0              0.0               0.0             0.000000   
98             0.0              0.0               0.0             0.000000   

    ...  Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0   ...         0.0                            0.0               0.0   
1   ...         0.0                            0.0               0.0   
2   ...         0.0                            0.0               0.0   
3   ...         0.0                            0.0               0.0   
4   ...         0.0                            0.0               0.0   
..  ...         ...                            ...               ...   
94  ...         0.0                            0.0               0.0   
95  ...         0.0                            0.0               0.0   
96  ...         0.0                            0.0               0.0   
97  ...         0.0                            0.0               0.0   
98  ...         0.0                            0.0               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0      0.000000               0.000000              0.0       0.0        0.0   
1      0.000000               0.000000              0.0       0.0        0.0   
2      0.000000               0.000000              0.0       0.0        0.0   
3      0.000000               0.000000              0.0       0.0        0.0   
4      0.000000               0.000000              0.0       0.0        0.0   
..          ...                    ...              ...       ...        ...   
94     0.000000               0.030303     

In [206]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

ating Rink  0.25
4  Modern European Restaurant  0.00


----Brockton, Parkdale Village, Exhibition Place----
                   venue  freq
0                   Café  0.12
1                 Bakery  0.08
2         Breakfast Spot  0.08
3  Performing Arts Venue  0.08
4            Coffee Shop  0.08


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  0.13
3          Boutique  0.07
4   Harbor / Marina  0.07


----Caledonia-Fairbanks----
                       venue  freq
0                       Park  0.50
1              Women's Store  0.25
2                        Bar  0.25
3  Middle Eastern Restaurant  0.00
4        Monument / Landmark  0.00


----Cedarbrae----
                  venue  freq
0       Thai Restaurant  0.12
1                Bakery  0.12
2           Gas Station  0.12
3  Caribbean Restaurant  0.12
4    Athletics & Sports

In [207]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [208]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1                        Pub                   Gym              Pharmacy   
2                       Bank        Sandwich Place            Restaurant   
3        Japanese Restaurant                  Bank    Chinese Restaurant   
4                Coffee Shop        Sandwich Place           Pizza Place   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0            Donut Shop                      Diner        Discount Store   
1        Sandwich Place                Coffee Shop          Skating Rink   
2        Ice Cream Shop  Middle Eastern Restaurant           Supermarket   
3         Women's Store                      Diner        Discount Store   
4       Thai Restaurant                  Juice Bar  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Distribution Center               Dog Run       Doner Restaurant  
1      Department Store          Dessert Shop     Dim Sum Restaurant  
2     Mobile Phone Shop          Intersection          Deli / Bodega  
3   Distribution Center               Dog Run       Doner Restaurant  
4               Butcher                  Café      Indian Restaurant

In [209]:
neighborhoods_venues_sorted.shape

(99, 11)

# Perform KMeans clustering.  Build 5 clusters

### I tried building 5, 6 and 7 clusters but the higher number of clusters generated multiple clusters that contained only one neighborhood

In [224]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the datafr
kmeans.labels_[0:10] 


array([1, 0, 1, 1, 1, 1, 1, 1, 1, 3])

In [226]:
# add clustering labels
del neighborhoods_venues_sorted['Cluster Labels']
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_metro

# merge toronto_grouped with toronto_metro to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Find neighborhoods without venues to use for comparison and drop them

In [227]:
print(toronto_merged.loc[~toronto_merged['Cluster Labels'].isin([0,1,2,3,4,5])])

   PostalCode      Borough              Neighborhood   Latitude  Longitude  \
5         M9A    Etobicoke          Islington Avenue  43.667856 -79.532242   
45        M2L   North York  York Mills, Silver Hills  43.757490 -79.374714   
52        M2M   North York   Willowdale, Newtonbrook  43.789053 -79.408493   
95        M1X  Scarborough               Upper Rouge  43.836125 -79.205636   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5              NaN                   NaN                   NaN   
45             NaN                   NaN                   NaN   
52             NaN                   NaN                   NaN   
95             NaN                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5                    NaN                   NaN                   NaN   
45                   NaN                   NaN                   NaN   
52                   NaN                   NaN                

### reset the cluster lables to integers 

In [229]:
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype(int)

toronto_merged.head() # check the last columns!

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.329656               1  Construction & Landscaping   
1 -79.315572               1                 Pizza Place   
2 -79.360636               1                 Coffee Shop   
3 -79.464763               1              Clothing Store   
4 -79.389494               1                 Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                    Park                  Pool      Food & Drink Shop   
1             Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                     Pub                Bakery                   Park   
3  Furniture / Home Store             Gift Shop           Carpet Store   
4        Sushi Restaurant           Yoga Studio                   Café   

  5th Most Common Venue        6th Most Common Venue  7th Most Common Venue  \
0      Doner Restaurant           Dim Sum Restaurant                  Diner   
1          Intersection  Eastern European Restaurant    Dumpling Restaurant   
2                  Café               Breakfast Spot                Theater   
3           Coffee Shop                     Boutique  Vietnamese Restaurant   
4    Italian Restaurant                          Bar                   Bank   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Discount Store   Distribution Center                Dog Run  
1     Electronics Store             Drugstore       Department Store  
2  Gym / Fitness Center        Farmers Market             Restaurant  
3     Accessories Store          Concert Hall           Dessert Shop  
4    Mexican Restaurant   Japanese Restaurant                  Diner

## Build map of Toronto with neighborhoods.  colored and labled by cluster

In [230]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.8).add_to(map_clusters)
       
map_clusters

# The following are the neighborhood clusters

In [245]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
#print("Pizza!")

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
50  North York               0           Pizza Place                   Gym   
63        York               0              Bus Line           Pizza Place   
77   Etobicoke               0           Pizza Place        Sandwich Place   
93   Etobicoke               0           Pizza Place                   Pub   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
50      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
63         Women's Store      Doner Restaurant    Dim Sum Restaurant   
77     Mobile Phone Shop              Bus Line   Distribution Center   
93                   Gym              Pharmacy        Sandwich Place   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
50                 Diner        Discount Store   Distribution Center   
63                 Diner        Discount Store   Distribution Center   
77    Dim Sum Restaurant                 Diner        Discount Store   
93           Coffee Shop          Skating Rink      Department Store   

   9th Most Common Venue 10th Most Common Venue  
50               Dog Run             Donut Shop  
63               Dog Run             Donut Shop  
77               Dog Run       Department Store  
93          Dessert Shop     Dim Sum Restaurant

## Cluster 0 I named Pizza!

In [233]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
#print("Shopping")

Borough  Cluster Labels       1st Most Common Venue  \
0               North York               1  Construction & Landscaping   
1               North York               1                 Pizza Place   
2         Downtown Toronto               1                 Coffee Shop   
3               North York               1              Clothing Store   
4             Queen's Park               1                 Coffee Shop   
..                     ...             ...                         ...   
97        Downtown Toronto               1                 Coffee Shop   
98               Etobicoke               1                       River   
99        Downtown Toronto               1                 Coffee Shop   
100  East Toronto Business               1          Light Rail Station   
102              Etobicoke               1                 Social Club   

      2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                      Park                  Pool      Food & Drink Shop   
1               Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                       Pub                Bakery                   Park   
3    Furniture / Home Store             Gift Shop           Carpet Store   
4          Sushi Restaurant           Yoga Studio                   Café   
..                      ...                   ...                    ...   
97                     Café                 Hotel             Restaurant   
98            Women's Store      Doner Restaurant     Dim Sum Restaurant   
99      Japanese Restaurant      Sushi Restaurant             Restaurant   
100    Gym / Fitness Center         Auto Workshop          Garden Center   
102       Convenience Store       Supplement Shop         Discount Store   

    5th Most Common Venue        6th Most Common Venue  7th Most Common Venue  \
0        Doner Restaurant           Dim Sum Restaurant                  Diner   
1            Intersection  Eastern European Restaurant    Dumpling Restaurant   
2                    Café               Breakfast Spot                Theater   
3             Coffee Shop                     Boutique  Vietnamese Restaurant   
4      Italian Restaurant                          Bar                   Bank   
..                    ...                          ...                    ...   
97    Japanese Restaurant                          Gym       Asian Restaurant   
98                  Diner               Discount Store    Distribution Center   
99            Yoga Studio         Fast Food Restaurant           Dance Studio   
100                Garden         Fast Food Restaurant         Farmers Market   
102         Tanning Salon                Burrito Place         Sandwich Place   

    8th Most Common Venue   9th Most Common Venue 10th Most Common Venue  
0          Discount Store     Distribution Center                Dog Run  
1       Electronics Store               Drugstore       Department Store  
2    Gym / Fitness Center          Farmers Market             Restaurant  
3       Accessories Store            Concert Hall           Dessert Shop  
4      Mexican Restaurant     Japanese Restaurant                  Diner  
..                    ...                     ...                    ...  
97     Seafood Restaurant              Steakhouse    American Restaurant  
98                Dog Run              Donut Shop       Department Store  
99                    Pub                   Hotel                Gay Bar  
100            Comic Shop                    Park       Recording Studio  
102          Burger Joint  Thrift / Vintage Store             Kids Store  

[83 rows x 12 columns]

## Cluster 1 -  I named "Shopping"

In [234]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
#print("International cuisine")

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke               2   Filipino Restaurant         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11            Donut Shop    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11        Discount Store   Distribution Center               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
11      Doner Restaurant              Drugstore

## cluster 2 -  this single neighborhood reminds me of international food

In [235]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
#print("Let's go to the park!")

Borough  Cluster Labels 1st Most Common Venue  \
21                     York               3                  Park   
35   East York/East Toronto               3          Intersection   
40               North York               3               Airport   
61          Central Toronto               3                  Park   
64                     York               3                  Park   
66               North York               3     Convenience Store   
68          Central Toronto               3         Jewelry Store   
85              Scarborough               3          Intersection   
91         Downtown Toronto               3                  Park   
101               Etobicoke               3        Baseball Field   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21          Women's Store                   Bar            Donut Shop   
35                   Park     Convenience Store         Women's Store   
40                   Park           Snack Place         Women's Store   
61               Bus Line           Swim School         Women's Store   
64          Women's Store            Donut Shop    Dim Sum Restaurant   
66                   Park            Donut Shop    Dim Sum Restaurant   
68                   Park      Sushi Restaurant                 Trail   
85             Playground                  Park      Doner Restaurant   
91                  Trail            Playground         Women's Store   
101         Deli / Bodega                  Park         Women's Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21                  Diner        Discount Store   Distribution Center   
35             Donut Shop                 Diner        Discount Store   
40             Donut Shop                 Diner        Discount Store   
61       Doner Restaurant                 Diner        Discount Store   
64                  Diner        Discount Store   Distribution Center   
66                  Diner        Discount Store   Distribution Center   
68       Doner Restaurant    Dim Sum Restaurant                 Diner   
85           Dessert Shop    Dim Sum Restaurant                 Diner   
91           Dessert Shop    Dim Sum Restaurant                 Diner   
101                 Diner        Discount Store   Distribution Center   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21                Dog Run      Doner Restaurant              Drugstore  
35    Distribution Center               Dog Run       Doner Restaurant  
40    Distribution Center               Dog Run       Doner Restaurant  
61    Distribution Center               Dog Run             Donut Shop  
64                Dog Run      Doner Restaurant              Drugstore  
66                Dog Run      Doner Restaurant          Women's Store  
68         Discount Store   Distribution Center                Dog Run  
85         Discount Store   Distribution Center                Dog Run  
91         Discount Store   Distribution Center                Dog Run  
101               Dog Run      Doner Restaurant             Donut Shop

## Cluster 3  -  is all about going to the park

In [236]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
#print("Take me out to the ball park")

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57  North York               4        Baseball Field         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57                 Diner        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
57             Drugstore         Farmers Market

## Cluster 4 - The highlight of this cluster is baseball!